File: Server/addMessage.php
<?php
  require "globalFunctions.php";
  $pdo = new PDO("pgsql:host=".HOST.";port=".PORT.";dbname=".DATABASE.";user=".USERNAME.";password=".PASSWORD);

  $myId = $_POST["myId"];
  $token = $_POST["token"];
  $buddyId = $_POST["buddyId"];
  $content = $_POST["content"];

	$isValid =
		isAuthenticated($myId, $token) &&
		isExtantUserId($buddyId) &&
		isValidString($content, 1, 1000);
	if (!$isValid) {
  	$pdo = null;
		die("Invalid");
	}

  $query = "select count(buddy_id) from (select distinct buddy_id from message where user_id = ? and date(sent_at) = current_date) as message;";
  $stmt = $pdo->prepare($query);
  $stmt->execute(array($myId));

  $row = $stmt->fetch();
  $dailyChatBuddies = $row[0];

	if ($dailyChatBuddies >= 50) {
 			$return = array(
				"dailyLimit" => 50,
				"isTooMany" => True,
			);
			echo json_encode($return);
			$pdo = null;
			exit;
	}

  $query = "insert into message (user_id, buddy_id, content) values (?, ?, ?) returning message_id;";
  $stmt = $pdo->prepare($query);
  $stmt->execute(array($myId, $buddyId, $content));

  $row = $stmt->fetch();
  $messageId = $row[0];

	sendNotification("message", $myId, $buddyId, $content);

  $return = array(
    "messageId" => $messageId
  );
  echo json_encode($return);

	$pdo = null;
?>

File: Server/addMessage.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

func addMessage(w http.ResponseWriter, r *http.Request) {
	r.ParseForm()
	token := r.Form.Get("token")
	buddyId := r.Form.Get("buddyId")
	content := r.Form.Get("content")
	myId := getMyId(token)

	isValid :=
		isAuthenticated(myId) &&
			isExtantUserId(buddyId) &&
			isValidString(content, 1, 256)
	if !isValid {
		fmt.Fprintln(w, "Invalid")
		return
	}

	var dailyChatBuddies int
	err := pool.QueryRow(context.Background(), "select count(buddy_id) from (select distinct buddy_id from message where user_id = $1 and date(sent_at) = current_date) as message;", myId).Scan(&dailyChatBuddies)
	handleError(err)

	if dailyChatBuddies >= 50 {
		mapper := map[string]interface{}{
			"dailyLimit": 50,
			"isTooMany":  true,
		}
		json, err := json.Marshal(mapper)
		handleError(err)

		w.Header().Set("Content-Type", "application/json")
		w.Write(json)
		return
	}

	var messageId string
	err = pool.QueryRow(context.Background(), "insert into message (user_id, buddy_id, content) values ($1, $2, $3) returning message_id;", myId, buddyId, content).Scan(&messageId)
	handleError(err)

	mapper := map[string]interface{}{
		"messageId": messageId,
	}
	json, err := json.Marshal(mapper)
	handleError(err)

	w.Header().Set("Content-Type", "application/json")
	w.Write(json)

	sendChannel(buddyId)
	sendNotification("message", myId, buddyId, content)
}
